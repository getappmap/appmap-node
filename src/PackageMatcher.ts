import { resolve } from "node:path";
import { fileURLToPath } from "node:url";

import fwdSlashPath from "./util/fwdSlashPath";
import { warn } from "./message";

export default class PackageMatcher extends Array<Package> {
  constructor(
    private root: string,
    packages: Package[],
  ) {
    // Normalize path separators
    packages.forEach((p) => {
      if (p.path) p.path = fwdSlashPath(p.path);
      if (p.exclude)
        for (let i = 0; i < p.exclude.length; i++) p.exclude[i] = fwdSlashPath(p.exclude[i]);
    });

    super(...packages);
    this.resolved = new Map(
      packages.filter((p) => p.path).map(({ path }) => [path!, fwdSlashPath(resolve(root, path!))]),
    );
  }

  private resolved: Map<string, string>;

  private resolve(path: string) {
    return this.resolved.get(path) ?? fwdSlashPath(resolve(this.root, path));
  }

  match(path: string): Package | undefined {
    if (path.startsWith("file:")) path = fileURLToPath(path);

    // Make sure passed path is forward slashed
    const fixedPath = fwdSlashPath(path);

    const pkg = this.find(
      (pkg) => pkg.path != undefined && fixedPath.startsWith(this.resolve(pkg.path)),
    );
    return pkg?.exclude?.find((ex) => fixedPath.includes(ex)) ? undefined : pkg;
  }

  matchLibrary(moduleId: string): Package | undefined {
    return this.find((pkg) => pkg.module?.replace(/^node:/, "") == moduleId.replace(/^node:/, ""));
  }
}

export interface Package {
  path?: string;
  exclude?: string[];
  module?: string;
  shallow?: boolean;
  prisma?: string; // custom prisma client module id
}

export function parsePackages(packages: unknown): Package[] | undefined {
  if (!packages || !Array.isArray(packages)) return;

  const result: Package[] = [];

  for (const pkg of packages as unknown[]) {
    if (typeof pkg === "string") result.push({ path: pkg });
    else if (typeof pkg === "object" && pkg !== null && ("path" in pkg || "module" in pkg)) {
      const entry: Package = {};
      if ("path" in pkg) entry.path = String(pkg.path);
      if ("exclude" in pkg) entry.exclude = Array.isArray(pkg.exclude) ? pkg.exclude : [];
      if ("module" in pkg && typeof pkg.module === "string") entry.module = pkg.module;
      if ("shallow" in pkg && typeof pkg.shallow === "boolean") entry.shallow = pkg.shallow;
      if ("prisma" in pkg && pkg.prisma != null && typeof pkg.prisma === "string")
        entry.prisma = pkg.prisma;

      if (entry.module && entry.path) {
        warn(
          "Package configuration cannot include 'path' if 'module' is specified. " +
            "Ignoring 'path' setting. ",
        );
        entry.path = undefined;
      }

      result.push(entry);
    }
  }

  if (result.length > 0) return result;
}
