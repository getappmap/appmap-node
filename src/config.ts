import assert from "node:assert";
import { readFileSync, writeFileSync } from "node:fs";
import { basename, join, resolve } from "node:path";
import { cwd } from "node:process";

import { PackageJson } from "type-fest";
import YAML, { Document } from "yaml";

import { info, warn } from "./message";
import PackageMatcher, { Package, parsePackages } from "./PackageMatcher";
import locateFileUp from "./util/findFileUp";
import lazyOpt from "./util/lazyOpt";
import tryOr from "./util/tryOr";

const responseBodyMaxLengthDefault = 10000;
const kResponseBodyMaxLengthEnvar = "APPMAP_RESPONSE_BODY_MAX_LENGTH";

export class Config {
  public readonly relativeAppmapDir: string;
  public readonly appName: string;
  public readonly root: string;
  public readonly configPath: string;
  public readonly default: boolean;
  public readonly packages: PackageMatcher;
  public readonly responseBodyMaxLength: number;
  public readonly language: string;

  private readonly document?: Document;
  private migrationPending = false;

  constructor(pwd = cwd()) {
    const configDir = locateFileUp("appmap.yml", process.env.APPMAP_ROOT ?? pwd);
    const packageDir = lazyOpt(() => locateFileUp("package.json", process.env.APPMAP_ROOT ?? pwd));
    const targetPackage = packageDir.then((dir) =>
      tryOr(() => JSON.parse(readFileSync(join(dir, "package.json"), "utf8")) as PackageJson),
    );

    const root = (this.root = process.env.APPMAP_ROOT ?? configDir ?? packageDir() ?? pwd);

    this.configPath = join(root, "appmap.yml");
    this.document = loadConfigDocument(this.configPath);
    const config = this.document ? readConfigFile(this.document) : undefined;
    this.default = !this.document;

    this.relativeAppmapDir = config?.appmap_dir ?? "tmp/appmap";

    this.appName = config?.name ?? targetPackage()?.name ?? basename(root);

    this.language = config?.language ?? "javascript";
    this.migrationPending ||= !!config && config.language === undefined;

    this.packages = new PackageMatcher(
      root,
      config?.packages ?? [
        {
          path: ".",
          exclude: ["node_modules", ".yarn"],
        },
      ],
    );

    let envResponseBodyMaxLength: number | undefined;
    if (process.env[kResponseBodyMaxLengthEnvar] != undefined) {
      const value = parseInt(process.env[kResponseBodyMaxLengthEnvar]);
      envResponseBodyMaxLength = value >= 0 ? value : undefined;
      if (envResponseBodyMaxLength == undefined)
        warn(`Environment variable ${kResponseBodyMaxLengthEnvar} must be a non-negative integer.`);
    }

    this.responseBodyMaxLength =
      envResponseBodyMaxLength ?? config?.response_body_max_length ?? responseBodyMaxLengthDefault;
  }

  private absoluteAppmapDir?: string;
  get appmapDir() {
    return (this.absoluteAppmapDir ||= resolve(this.root, this.relativeAppmapDir));
  }

  get prismaClientModuleIds(): string[] {
    const result = ["@prisma/client"];
    this.packages.forEach((p) => {
      if (p.prisma) result.push(p.prisma);
    });
    return result;
  }

  private packageMap = new Map<string, Package | undefined>();

  getPackage(pathOrModuleId: string | undefined, isLibrary: boolean) {
    if (!pathOrModuleId) return;
    const key = `${pathOrModuleId}:${isLibrary}`;

    if (!this.packageMap.has(key)) {
      const pkg = isLibrary
        ? this.packages.matchLibrary(pathOrModuleId)
        : this.packages.match(pathOrModuleId);

      this.packageMap.set(key, pkg);
    }
    return this.packageMap.get(key);
  }

  getFunctionLabels(
    pkg: Package | undefined,
    functionName: string,
    klass?: string,
  ): string[] | undefined {
    if (!pkg?.functions) return;
    return pkg.functions
      .filter(
        (group) =>
          group.names?.includes(functionName) ??
          (klass != undefined && group.names.includes(klass + "." + functionName)),
      )
      .flatMap((group) => group.labels);
  }

  export() {
    process.env.APPMAP_ROOT = this.root;
  }

  toJSON(): ConfigFile {
    return {
      name: this.appName,
      language: this.language,
      appmap_dir: this.relativeAppmapDir,
      packages: this.packages,
    };
  }

  migrate() {
    if (!this.migrationPending) return;

    if (this.document) {
      this.document.set("language", this.language);
    }

    info("appmap.yml requires migration, changes will be automatically applied.");
    writeFileSync(
      this.configPath,
      YAML.stringify(this.document ?? this, { keepSourceTokens: true }),
    );
  }
}

interface ConfigFile {
  appmap_dir?: string;
  name?: string;
  packages?: Package[];
  response_body_max_length?: number;
  language?: string;
}

// Maintaining the YAML document is important to preserve existing comments and formatting
// in the original file.
function loadConfigDocument(path: string | undefined): Document | undefined {
  if (!path) return;

  let fileContent: string;
  try {
    fileContent = readFileSync(path, "utf-8");
  } catch (exn) {
    if (exn && typeof exn == "object" && "code" in exn && exn.code == "ENOENT") return;
    throw exn;
  }

  const document = YAML.parseDocument(fileContent, { keepSourceTokens: true });
  if (document.errors.length > 0) {
    const errorMessage = document.errors.map((e) => `${e.name}: ${e.message}`).join("\n");
    throw new Error(
      `Failed to parse config file at ${path}\n${errorMessage}\nYou can remove the file to use the default configuration.`,
    );
  }
  return document;
}

function readConfigFile(document: Document): ConfigFile {
  const config = document.toJSON() as Record<string, unknown>;
  const result: ConfigFile = {};
  assert(typeof config === "object");
  if ("name" in config) result.name = String(config.name);
  if ("appmap_dir" in config) result.appmap_dir = String(config.appmap_dir);
  if ("packages" in config) result.packages = parsePackages(config.packages);
  if ("language" in config) result.language = String(config.language);
  if ("response_body_max_length" in config) {
    const value = parseInt(String(config.response_body_max_length));
    result.response_body_max_length = value >= 0 ? value : undefined;
  }

  return result;
}

export default new Config();
