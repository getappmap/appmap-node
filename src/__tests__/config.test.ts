import { mkdirSync, readFileSync, writeFileSync } from "node:fs";
import { basename } from "node:path";
import { chdir, cwd } from "node:process";

import tmp from "tmp";
import { PackageJson } from "type-fest";
import YAML from "yaml";

import PackageMatcher from "../PackageMatcher";
import { Config } from "../config";
import { fixAbsPath } from "../hooks/__tests__/fixAbsPath";

tmp.setGracefulCleanup();

// https://github.com/nodejs/node/issues/11422
// On MacOS, temp paths can be reported as '/var/folders/...'
// or '/private/var/folders/...' based on the API call.
// One is symlink to the other.
// This fix prevents needless test failures on local MacOS.
function fixMacOsTmpPath(path: string) {
  if (path?.startsWith("/var/folders")) return `/private${path}`;
  return path;
}

describe(Config, () => {
  it("respects environment variables", () => {
    process.env.APPMAP_ROOT = "/test/app";
    const config = new Config();
    expect(config).toMatchObject({
      root: "/test/app",
      relativeAppmapDir: "tmp/appmap",
      appName: "app",
      packages: new PackageMatcher("/test/app", [
        { path: ".", exclude: ["node_modules", ".yarn"] },
      ]),
    });
  });

  it("uses the current directory for root", () => {
    expect(new Config()).toMatchObject({
      root: fixMacOsTmpPath(dir),
      relativeAppmapDir: "tmp/appmap",
      appName: basename(dir),
    });
  });

  it("searches for package.json and uses package name from it", () => {
    writeFileSync("package.json", JSON.stringify({ name: "test-package" } as PackageJson));

    mkdirSync("subdirectory");
    chdir("subdirectory");
    expect(new Config()).toMatchObject({
      root: fixMacOsTmpPath(dir),
      relativeAppmapDir: "tmp/appmap",
      appName: "test-package",
    });
  });

  it("searches for appmap.yml and uses config from it", () => {
    writeFileSync(
      "appmap.yml",
      YAML.stringify({
        name: "test-package",
        appmap_dir: "appmap",
        packages: [{ path: ".", exclude: ["excluded"] }, "../lib"],
      }),
    );

    mkdirSync("subdirectory");
    chdir("subdirectory");
    expect(new Config()).toMatchObject({
      root: fixMacOsTmpPath(dir),
      relativeAppmapDir: "appmap",
      appName: "test-package",
      packages: new PackageMatcher(fixMacOsTmpPath(dir), [
        { path: ".", exclude: ["excluded"] },
        { path: "../lib" },
      ]),
    });
  });

  it("uses default packages if the field in appmap.yml has unrecognized format", () => {
    writeFileSync(
      "appmap.yml",
      YAML.stringify({
        name: "test-package",
        appmap_dir: "appmap",
        packages: [{ regexp: "foo", enabled: false }],
      }),
    );

    mkdirSync("subdirectory");
    chdir("subdirectory");
    expect(new Config()).toMatchObject({
      root: fixMacOsTmpPath(dir),
      relativeAppmapDir: "appmap",
      appName: "test-package",
      packages: new PackageMatcher(fixMacOsTmpPath(dir), [
        { path: ".", exclude: ["node_modules", ".yarn"] },
      ]),
    });
  });

  it("if the appmap.yml is malformed throws with an informative message", () => {
    const malformedAppMapFileContent = "{ appmap-dir: 'abc/xyz' ";
    writeFileSync("appmap.yml", malformedAppMapFileContent);

    expect(() => new Config()).toThrowError(
      /You can remove the file to use the default configuration\.$/,
    );

    expect(readFileSync("appmap.yml").toString()).toEqual(malformedAppMapFileContent);
  });

  let dir: string;
  beforeEach(() => {
    chdir((dir = tmp.dirSync().name));
    jest.replaceProperty(process, "env", {});
  });

  const origCwd = cwd();
  afterEach(() => chdir(origCwd));
});

describe(PackageMatcher, () => {
  it("matches packages", () => {
    const pkg = { path: ".", exclude: ["node_modules", ".yarn"] };
    const matcher = new PackageMatcher(fixAbsPath("/test/app"), [pkg]);
    expect(matcher.match(fixAbsPath("/test/app/lib/foo.js"))).toEqual(pkg);
    if (process.platform == "win32")
      expect(matcher.match(fixAbsPath("\\test\\app\\lib\\foo.js"))).toEqual(pkg);
    expect(matcher.match(fixAbsPath("/other/app/lib/foo.js"))).toBeUndefined();
    expect(matcher.match(fixAbsPath("/test/app/node_modules/lib/foo.js"))).toBeUndefined();
    expect(matcher.match(fixAbsPath("/test/app/.yarn/lib/foo.js"))).toBeUndefined();
  });
});
