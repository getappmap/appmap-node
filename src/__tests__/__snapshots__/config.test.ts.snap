// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Config migrate does not update an existing config if the language field is already set 1`] = `
"name: test-package
language: ruby
appmap_dir: appmap
packages:
  - path: .
"
`;

exports[`Config migrate sets the language field in existing configs 1`] = `
"name: test-package
appmap_dir: appmap
packages:
  - path: .
language: javascript
"
`;
